### DISCUSSION TITLE
How do you pass "more data" to collection "Item" component?

### DISCUSSION THREAD

**Discussion by KasTasElis** [adobe/react-spectrum#6373] [Q&A]
I am building a custom Design System using react-aria and react-stately hooks. I am confused how ["Collections"](https://react-spectrum.adobe.com/react-stately/collections.html) work. I have read the documentation a couple of times, looked at examples and [architecture docs](https://react-spectrum.adobe.com/architecture.html), I am still confused... 

Here's an example of a problem I am struggling with while building a custom Tab Navigation component. 

In react stately docs it explains that [you should use the "Item" component to only "provide data to the parent collection component"](https://react-spectrum.adobe.com/react-stately/collections.html#static-collections), the parent component should take care of the rendering, layout, styling etc... 

Heres a [simple working example](https://codesandbox.io/s/practical-monad-punzo?file=/src/Tabs.js) of what they provide. 

However, my custom Tab Navigation component requires nav buttons to have an optional Icon and a sub-title... 

I am confused, how do i pass this additional data? Via the "Item's" component props? TypeScript starts shouting at me for adding unspecified additional props to "Item" if i attempt to do like below...? 

```
<Tabs>
    <Item title="One" subTitle="Terms & Conditions" icon="Hello">
        Panel content goes here! - Lorem Ipsum has been the industrys standard dummy text ever since
        the 1500s, when an unknown printer took a galley of type and scrambled it to make a type
        specimen book. It has survived no
    </Item>
    <Item subTitle="Terms & Conditions" title="Two" icon="Hello">
        Panel content goes here again! - Lorem Ipsum has been the industrys standard dummy text ever
        since the 1500s, when an unknown printer took a galley of type and scrambled it to make a
        type specimen book. It has survived no
    </Item>
</Tabs>
```

```
Type '{ children: Element; subTitle: string; title: string; icon: string; }' is not assignable to type 'IntrinsicAttributes & ItemProps<unknown>'.
  Property 'subTitle' does not exist on type 'IntrinsicAttributes & ItemProps<unknown>'.ts(2322)
```

What am i missing / not understanding here? 🤔

Am i supposed to create some kind of a wrapper component for Item here with the additional props? 🤔 

Like so? 

```
const ItemWrapper = ({ title, subTitle, icon, children, ...props }) => {
  return (
    <Item
      key={title}
      title={
        <div>
          {icon ? <Icon name={icon} /> : null}
          <div>{title}</div>
          {subTitle ? <div>{subTitle}</div> : null}
        </div>
      }
      {...props}
    >
      {children}
    </Item>
  );
};

export const MyTabs = () => (
  <Tabs>
    <ItemWrapper title="One" subTitle="Terms & Conditions" icon="Hello">
      <div>
        Panel content goes here! - Lorem Ipsum has been the industrys standard dummy text ever since
        the 1500s, when an unknown printer took a galley of type and scrambled it to make a type
        specimen book. It has survived no
      </div>
    </ItemWrapper>
    <ItemWrapper subTitle="Terms & Conditions" title="Two" icon="Hello">
      <div>
        Panel content goes here again! - Lorem Ipsum has been the industrys standard dummy text ever
        since the 1500s, when an unknown printer took a galley of type and scrambled it to make a
        type specimen book. It has survived no
      </div>
    </ItemWrapper>
  </Tabs>
);
```

But doing it this way is no longer just "providing data to item" is it not? I am giving a layout now... which is a rendering concern... 

What would be the correct way of doing this? 🤔

Thank you.

**Comment by ktabors** [comment]
Your `<ItemWrapper />` idea won't work because you can't wrap an `<Item />`. Your wrapper idea does identify that the `title` prop takes a React node and you can pass in the icon, title, and subtitle that way. To see the full list of Item props you can click on "Item" -> "ItemProps" on the React Stately Collections page, see the following image for where it is on the page.
<img width="241" alt="image" src="https://github.com/adobe/react-spectrum/assets/309540/3612f5f2-62a2-4d1d-b711-3b042960b602">

You should consider looking at the [Tabs](https://react-spectrum.adobe.com/react-aria/Tabs.html) in React Aria Components and read the React Aria [Collections](https://react-spectrum.adobe.com/react-aria/collections.html) documentation. They are easier to use and would allow you to wrap an `Item`. You can also see in the default [example](https://react-spectrum.adobe.com/react-aria/Tabs.html#example) that it doesn't use `Item` or the `title` prop and you could have your icon, title, and subtitle as children to a Tab.

If you're only wanting to use the hooks, you could use an inline solution by creating a `TabItemTitle` component that looks like the following code block (add your conditional logic) and would be passed to the `title` prop of an Item like: `<Item title={<TabItemTitle icon={<Star />} title="Recent" subtitle="Subtitle" />}>...</Item>`. `<Star />` is just a wrapper around an SVG to represent an icon.
```
function TabItemTitle(props) {
  return (
    <>
      <span className="icon">
        <Star />
      </span>
      <span className="title">Recent</span>
      <span className="subtitle">Subtitle</span>
    </>
  );
}
```

**Reply by KasTasElis** [reply]
Hello there, thanks for answering! 🙌

I have read both [Collection](https://react-spectrum.adobe.com/react-aria/collections.html) and [Tabs](https://react-spectrum.adobe.com/react-aria/Tabs.html) documentation before asking this question, I ended up refactoring to use JSX Component API instead of Hooks as a solution for now.

For Hooks implementation, passing a JSX structure to Item's title property feels odd because in docs it is stated that this should not be the concern of an Item. But then why allow the title prop take a ReactNode 🤔... 

Feels like the better thing to do would be to allow Item to take additional props, let the user pass any additional data they need for rendering, not a JSX structure.

**Reply by ktabors** [reply]
Thanks for the feedback. We're happy you were able to use RAC Tabs. I see your point about the collection documentation suggesting the data shouldn't define the rendering and that being at odds with my suggestion. This is a special case as far as our collection components go. The item has both the Tab title and the Tab content. It would be better if the documentation example were more complicated and followed the [link](https://react-spectrum.adobe.com/react-aria/useTabList.html#links) example at the bottom of the useTabList page or RAC Tabs which use a TabList and TabPanel to separate these two parts of a Tab into distinct items.
